Predictive variance acquisition

satrialoka opened the pull request:
Add predictive variance acquisition function for Active Learning
a continuation of #185

Feature Implemented:
PredictiveVariance acquisition function

Follow up:
After this PR, the bayesian active learning by disagreement (BALD) will be implemented

- [x] update notebooks
- [x] add ref

this will close #185

satrialoka committed:
wip

satrialoka committed:
wip

satrialoka committed:
Merge branch 'develop' of https://github.com/secondmind-labs/trieste into predictive_variance_acquisition

satrialoka committed:
use log-branin

satrialoka committed:
add predvar

joelberkeley commented on trieste/acquisition/function.py:
```diff
@@ -1365,3 +1365,29 @@ def repulsion_term(
+        """
+        return lambda at: self.acquisition(model, at)
+
+    @staticmethod
+    def acquisition(model: ProbabilisticModel, at: TensorType) -> TensorType:
+        _, covariance = model.predict_joint(at)
+        return tf.linalg.det(covariance)
```
any specific need for this public method? this could be a function in `prepare_acquisition_function`

satrialoka replied:
no actually, yeah that'll be neat to make it a function inside prepare_acqusition_function

satrialoka replied:
correct me if I am wrong, but making this public method is needed for test code no?
for example in [negative lcb](https://github.com/secondmind-labs/trieste/blob/4ef8d1317defca77b8f0dceed9ccefee894daf8f/tests/unit/acquisition/test_function.py#L375) test. so it should be a function **outside** of PredVar class. WDYT?

joelberkeley replied:
when you call `prepare...` you essentially get `determinantcovariance` back immediately, so you can test via `prepare...`, with  a stub model with custom `predict_joint`
```
class Model(ProbabilisticModel):
    def predict_joint(x):
        mean = <some function of x that helps you know this isn't being used>
        cov = <some function of x that makes it easy to test>
        return mean, cov
```
and use any old data (you could use some data that gives confidence it's not being used, but I'm not sure that's necessary).

i'm a big fan of keeping as much out of the public API as possible. It helps keep the intention of the API clear. In some cases, we exposed the underlying acquisition function as well as the `AcquisitionFunctionBuilder`. That might be because users might not want to do the work the builder does in the way the builder does it (EI can use different "best" points, for example), or it could be for ease of use (people are perhaps more used to thinking about LCB than negative LCB). In some cases, it may have just been a mistake. I hope we never did it to help with testing. I don't think it's good practice to make things public solely for testing purposes

satrialoka replied:
Ah yeah, that makes sense. yeah, I think that'll be the best approach here. Thanks for the suggestion!

satrialoka committed:
move acquisition inside

satrialoka committed:
wip notebook

satrialoka committed:
wip

satrialoka committed:
add notebook to tutorial.rst

satrialoka committed:
convert to to percentage script

joelberkeley commented on trieste/acquisition/function.py:
```diff
@@ -1365,3 +1365,29 @@ def repulsion_term(
+        :return: The determinant of the predictive function.
+        """
+
+        def determinantcovariance(model: ProbabilisticModel, at: TensorType) -> TensorType:
+            _, covariance = model.predict_joint(at)
+            return tf.linalg.det(covariance)
+
+        return lambda at: determinantcovariance(model, at)
```
you can use `model` from the larger scope (indeed you're already doing this in the `lambda`)
```suggestion
        def determinantcovariance(at: TensorType) -> TensorType:
            _, covariance = model.predict_joint(at)
            return tf.linalg.det(covariance)

        return determinantcovariance
```

satrialoka committed:
Update trieste/acquisition/function.py

Co-authored-by: Joel Berkeley <16429957+joelberkeley@users.noreply.github.com>

satrialoka committed:
predictive variance builder test

satrialoka committed:
Merge branch 'predictive_variance_acquisition' of https://github.com/satrialoka/trieste into predictive_variance_acquisition

satrialoka committed:
format

satrialoka committed:
Merge branch 'develop' into predictive_variance_acquisition

satrialoka committed:
test predictivevariance acq shape

satrialoka committed:
use bo history instead

satrialoka committed:
typo fixes

satrialoka committed:
Merge branch 'develop' into predictive_variance_acquisition

henrymoss commented on docs/notebooks/active_learning.pct.py:
```diff
@@ -0,0 +1,177 @@
+# # Active Learning
+
+# %% [markdown]
+# Sometimes rather than optimizing a black-box function, we want to learn the function better by reducing our model uncertainty. The one way to do this is by doing sampling in unknown inputs regions (active learning). This notebook demonstrates to perform Bayesian active learning using `trieste`.
```
Sometimes, we may just want to learn a black-box function, rather than optimizing it.  This goal is known as active learning and corresponds to choosing query points that reduce our model uncertainty. This notebook demonstrates to perform Bayesian active learning using `trieste`.

henrymoss commented on docs/notebooks/active_learning.pct.py:
```diff
@@ -0,0 +1,177 @@
+# %% [markdown]
+# ## Describe the problem
+#
+# In this example, we will consider doing active learning for the log branin function.
```
In this example, we will perform active learning for the log Branin function

henrymoss commented on docs/notebooks/active_learning.pct.py:
```diff
@@ -0,0 +1,177 @@
+fig.show()
+
+# %% [markdown]
+# We begin our Bayesian active learning steps after evaluating two initial points sampled by Halton sequence.
```
We begin our Bayesian active learning from a two-point initial design built from a space-filling Halton sequence.

henrymoss commented on docs/notebooks/active_learning.pct.py:
```diff
@@ -0,0 +1,177 @@
+# %% [markdown]
+# ## Active learning using predictive variance
+#
+# To do our first active learning example, we will use `PredictiveVariance`, one of the simplest active learning acquisition function. the `PredictiveVariance` sample a point on highest determinant of the predictive covariance of our model. Here, we will consider single query point using `PredictiveVariance`, the batch case will be shown later. We can utilize trieste's `BayesianOptimizer` to do t ...
```
For our first active learning example, we will use a simple acquisition function known as `PredictiveVariance` which chooses points for which we are highly uncertain (i.e. the predictive posterior covariance matrix at these points has large determinant).

We will now demonstrate how to choose individual query points using `PredictiveVariance` before moving onto batch active learning. For both cases, we can utilize trieste's `BayesianOptimizer` to do the active learning steps.

henrymoss commented on docs/notebooks/active_learning.pct.py:
```diff
@@ -0,0 +1,177 @@
+model = build_model(initial_data)
+
+# %% [markdown]
+# ## Active learning using predictive variance
```
Can you add a ref here?

satrialoka replied:
Couldnt find any paper that focuses on predictive variance
one candidate is this paper https://arxiv.org/pdf/0805.4359.pdf on chapter 2.2.1, they mention this a litle bit.
WDYT?

satrialoka replied:
@henrymoss  any comment on this?

vpicheny replied:
Possibly it's simple enough that we don't need a ref. In the case of batch > 1, I don't think that using the determinant of the covmatrix is standard and we could use a reference.

vpicheny replied:
maybe use 
MacKay, D. J. (1992). Information-based objective functions for active data selection. Neural computation, 4(4), 590-604.
instead?

hstojic replied:
I don't see the MacKay reference here in the text? can you please add it here as well, not just in `function.py`?

satrialoka replied:
okay 👌🏼

henrymoss commented on docs/notebooks/active_learning.pct.py:
```diff
@@ -0,0 +1,177 @@
+observations = dataset.observations.numpy()
+
+# %% [markdown]
+# Finally, we can check the performance of our `PredictiveVariance` active learning acquisition function by plotting the predictive variance landscape of our model. We can see how it samples on regions with high variance value.
```
Finally, we can check the performance of our `PredictiveVariance` active learning acquisition function by plotting the predictive variance landscape of our model. We can see how it samples regions for which our model is highly uncertain.

henrymoss commented on docs/notebooks/active_learning.pct.py:
```diff
@@ -0,0 +1,177 @@
+# %% [markdown]
+# ## Batch active learning using predictive variance
+#
+# For batch active learning using `PredictiveVariance`, we can set the `num_query_points` with the size of batch we want on the `EfficientGLobalOptimization` rule. The rest step is the same with the single query point of `PredictiveVariance` acquisition function.
```
For batch active learning, we must pass a `num_query_points` input to our `EfficientGLobalOptimization` rule.

henrymoss commented on docs/notebooks/active_learning.pct.py:
```diff
@@ -0,0 +1,177 @@
+dataset = result.try_get_final_dataset()
+
+query_points = dataset.query_points.numpy()
+observations = dataset.observations.numpy()
+
+from util.plotting import plot_bo_points, plot_function_2d
+
+for i in range(bo_iter):
+
```
Need to describe this plot

henrymoss commented on trieste/acquisition/function.py:
```diff
@@ -1365,3 +1365,29 @@ def repulsion_term(
+        :return: The determinant of the predictive function.
+        """
+
+        def determinantcovariance(at: TensorType) -> TensorType:
+            _, covariance = model.predict_joint(at)
+            return tf.linalg.det(covariance)
+
```
I wonder if its worth being able to specify an optional jitter param. Also, might be worth returning log(det), as this can be calculated much more numerically stably in tensorflow (https://www.tensorflow.org/api_docs/python/tf/linalg/logdet) .

satrialoka replied:
yeah that'll be nice I think. I'll add it.

satrialoka replied:
got cholesky decomposition error when using `tf.linalg.logdet` for batch case. I'll investigate why this happens

henrymoss replied:
When does the cholesky error pop up? Do you remember to exp afterwards?

satrialoka replied:
ah yeah seems i forgot to exp

henrymoss reviewed (commented):
Have you thought about supporting greedy batch design for this acq function?

satrialoka committed:
wording and typos.

satrialoka committed:
Merge branch 'predictive_variance_acquisition' of https://github.com/satrialoka/trieste into predictive_variance_acquisition

satrialoka committed:
add custom jitter

satrialoka commented:
> 
> 
> Have you thought about supporting greedy batch design for this acq function?

haven't thought about it, do you think it is necessary?

satrialoka committed:
Merge branch 'develop' into predictive_variance_acquisition

henrymoss commented:
Don't bother with that for now

satrialoka committed:
using tf logdet

satrialoka committed:
Merge branch 'predictive_variance_acquisition' of https://github.com/satrialoka/trieste into predictive_variance_acquisition

satrialoka committed:
Merge branch 'develop' into predictive_variance_acquisition

satrialoka committed:
type

satrialoka committed:
update branin import

satrialoka committed:
mk_observer moved

satrialoka committed:
Merge branch 'develop' into predictive_variance_acquisition

satrialoka committed:
scaled branin and GPflowModelConfig

satrialoka committed:
add reference for predictive variance acq

vpicheny commented on docs/notebooks/active_learning.pct.py:
```diff
@@ -0,0 +1,188 @@
+# # Active Learning
+
+# %% [markdown]
+# Sometimes, we may just want to learn a black-box function, rather than optimizing it. This goal is known as active learning and corresponds to choosing query points that reduce our model uncertainty. This notebook demonstrates to perform Bayesian active learning using `trieste`.
```
how to perform

satrialoka committed:
change ref to Mackay1992

satrialoka committed:
missing word fix

satrialoka committed:
Merge branch 'develop' of https://github.com/secondmind-labs/trieste into predictive_variance_acquisition

satrialoka committed:
remove gramacy 2009 citation

henrymoss committed:
Merge branch 'develop' into predictive_variance_acquisition

apaleyes commented on trieste/acquisition/function.py:
```diff
@@ -1963,3 +1963,30 @@ def _get_min_value_samples(
+    ) -> AcquisitionFunction:
+        """
+        :param dataset: Unused.
+        :param model: The model over the specified ``dataset``.
+        :return: The determinant of the predictive function.
```
`jitter` is not in the docstring

apaleyes commented on docs/notebooks/active_learning.pct.py:
```diff
@@ -0,0 +1,188 @@
+        contour=True,
+        colorbar=True,
+        figsize=(10, 6),
+        title=["Variance contour with queried points at iter:" + str(i + 1)],
```
```suggestion
        title=["Variance contour with queried points at iter: " + str(i + 1)],
```

apaleyes commented on docs/notebooks/active_learning.pct.py:
```diff
@@ -0,0 +1,188 @@
+# %% [markdown]
+# ## Batch active learning using predictive variance
+#
+# For batch active learning, we must pass a num_query_points input to our `EfficientGLobalOptimization` rule.
```
```suggestion
# For batch active learning, we must pass a num_query_points input to our `EfficientGlobalOptimization` rule.
```

satrialoka committed:
add jitter docstring

satrialoka committed:
Update docs/notebooks/active_learning.pct.py

Co-authored-by: Andrei Paleyes <apaleyes@users.noreply.github.com>

satrialoka committed:
Update docs/notebooks/active_learning.pct.py

Co-authored-by: Andrei Paleyes <apaleyes@users.noreply.github.com>

satrialoka committed:
Merge branch 'predictive_variance_acquisition' of https://github.com/satrialoka/trieste into predictive_variance_acquisition

hstojic commented on trieste/acquisition/function.py:
```diff
@@ -1963,3 +1963,32 @@ def _get_min_value_samples(
+        :return: The determinant of the predictive function.
+        """
+
+        def determinantcovariance(at: TensorType) -> TensorType:
```
ideally, here you would use the `@tf.function` decorator, right @uri-granta?

for consistency with the rest of the acq functions codebase, could you please move out this function outside of the class, similar to say `lower_confidence_bound` function - this would also facilitate testing, no need to use a builder, so you would test separately builder and the function itself

hstojic commented on trieste/acquisition/function.py:
```diff
@@ -1963,3 +1963,32 @@ def _get_min_value_samples(
+            _, covariance = model.predict_joint(at)
+            return tf.exp(tf.linalg.logdet(covariance + jitter))
+
+        return determinantcovariance
```
I think it might be good to define `update_acquisition_function` and return `function` to avoid `prepare_acquisition_function` call

uri-granta replied:
Making the acquisition functions support in pace updates is not always trivial (and I don't think we should insist on it for every PR, as we can always add it later). However, in this case there isn't anything that needs updating, so you can have a trivial update_acquisition_function like in NegativeLowerConfidenceBound.

That said, you should probably move the jitter parameter to the initialiser (like in BatchMonteCarloExpectedImprovement) rather than in prepare_acqusition_function, as it doesn't belong in that function signature.

Hope that makes sense!

hstojic replied:
+1 for moving out jitter argument

hstojic commented on trieste/acquisition/function.py:
```diff
@@ -1963,3 +1963,32 @@ def _get_min_value_samples(
+            the covariance matrix.
+        :return: The determinant of the predictive function.
+        """
+
```
it would be good to assert that the model has `predict_joint` method and return a user-friendly message, as only some models will have this method implemented

hstojic commented on trieste/acquisition/function.py:
```diff
@@ -1963,3 +1963,32 @@ def _get_min_value_samples(
+
+class PredictiveVariance(SingleModelAcquisitionBuilder):
+    """
+    Builder for the determinant of the predictive covariance matrix over the batch points.
```
please add a note that the model needs to be able to supply covariance of the joint marginal distribution (which is also quite expensive to compute)

hstojic commented on tests/unit/acquisition/test_function.py:
```diff
@@ -1726,3 +1727,33 @@ def acquisition(x: TensorType) -> TensorType:
+    _, covariance = model.predict_joint(query_at)
+    expected = tf.linalg.det(covariance)
+    npt.assert_array_almost_equal(acq_fn(query_at), expected)
+
```
please add a test for raising an exception if the model doesn't implement `predict_joint`, e.g. GPflux models do not implement it, even though there is a method (required by the `ProbabilisticModel`

hstojic replied:
also please add `PredictiveVariance` to the `test_single_model_acquisition_function_builder_reprs`

hstojic replied:
next, we would need a test that it gives correct values, not that it just builds an acquisition function, see e.g. `test_expected_improvement` or `test_lower_confidence_bound`

hstojic replied:
finally, if you are willing to delve with `tf.function`, we would need smth like `test_negative_lower_confidence_bound_builder_updates_without_retracing`, I think copying this with minimal modifications should do the trick

satrialoka replied:
> please add a test for raising an exception if the model doesn't implement `predict_joint`, e.g. GPflux models do not implement it, even though there is a method required by the `ProbabilisticModel`. 

I was using this:

       if not hasattr(model, "predict_joint"):
            raise AttributeError(
                """
                PredictiveVariance only supports models with a predict_joint method.
                """
            )

but it doesn't pass the raise test since the `predicte_joint ` indeed is there in the base class. another option is to try-catch inside the function(not the builder), but it seems to do this inside the builder is much preferable, what are our options here?

satrialoka replied:
maybe @uri-granta  have any suggestion?

hstojic replied:
yes, `ProbabilisticModel` will always have the method, but those models that do not have it should keep the `raise NotImplementedError` so try checking that

satrialoka replied:
This is exactly i don't know how to do, 
some alternatives are checking the method's code by using  `from inspect import getsourcelines` to check whether there is a `NotImplementedError` inside it or not, but importing new lib like this seems not a neat way to do it.

hstojic replied:
Ok, I can chip in here with this test - can you please resolve all other comments in the meantime?

satrialoka replied:
okay 👌🏼

hstojic replied:
great, let me know when you are done!

hstojic replied:
ok, done

hstojic commented on docs/notebooks/active_learning.pct.py:
```diff
@@ -0,0 +1,188 @@
+    gpr = gpflow.models.GPR(data.astuple(), kernel, noise_variance=1e-5)
+    gpflow.set_trainable(gpr.likelihood, False)
+
+    return GPflowModelConfig(
```
Lets simplify this and use `GaussianProcessRegression` wrapper directly instead of `GPflowModelConfig`

hstojic commented on docs/notebooks/active_learning.pct.py:
```diff
@@ -0,0 +1,188 @@
+# %% [markdown]
+# ## Active learning using predictive variance
+#
+# For our first active learning example, we will use a simple acquisition function known as `PredictiveVariance` which chooses points for which we are highly uncertain (i.e. the predictive posterior covariance matrix at these points has large determinant).
```
can you please add a note that a model needs to have `predict_joint` method, i.e. be able to return full covariance - not all models will have that and likely to be expensive

hstojic commented on docs/notebooks/active_learning.pct.py:
```diff
@@ -0,0 +1,188 @@
+# %%
+from util.plotting import plot_bo_points, plot_function_2d
+
+for i in range(bo_iter):
```
you could make this a function, since you use exactly the same code below to plot batch results

hstojic commented on docs/notebooks/active_learning.pct.py:
```diff
@@ -0,0 +1,188 @@
+
+
+# %% [markdown]
+# Now we can visualize the batch predictive variance using our plotting function.
```
perhaps you could mention what kind of differences you could expect with using the batch method, say if you use same number of points above and total points here

hstojic reviewed (commented):
I have left some suggestions for improvements, let me know if you have any questions

satrialoka committed:
Merge branch 'develop' into predictive_variance_acquisition

satrialoka committed:
}

satrialoka committed:
restructure code

satrialoka committed:
only support model with predict_joint

satrialoka committed:
add notes on covariance

hstojic commented on trieste/acquisition/function.py:
```diff
@@ -2052,26 +2052,62 @@ class PredictiveVariance(SingleModelAcquisitionBuilder):
-            _, covariance = model.predict_joint(at)
-            return tf.exp(tf.linalg.logdet(covariance + jitter))
+        pv = determinantcovariance(model, self._jitter)
+        return tf.function(lambda at: pv(at))
```
you don't need another `tf.function` here I think, right @uri-granta?

uri-granta replied:
That's right: you can just return `determinantcovariance(model, self._jitter)` directly: no need to wrap it up in a lambda (unlike NegativeLowerConfidenceBound, which is negating the lcb).

satrialoka replied:
I see, okay 👌🏼

satrialoka committed:
black

satrialoka committed:
add to test_reprs

satrialoka committed:
missing .

satrialoka committed:
Use GaussianProcessRegresssion instead

satrialoka committed:
note for predic_joint method

satrialoka committed:
plotting function and set initial point to 4

satrialoka committed:
add citation in notebook

satrialoka committed:
remove tf.function() return

satrialoka committed:
test_predictive_variance

satrialoka committed:
test_retracing

satrialoka committed:
Merge branch 'develop' into predictive_variance_acquisition

satrialoka committed:
add more text in notebook regarding batch

hstojic committed:
added test for raising

hstojic committed:
formatting

hstojic committed:
fixed test; formatting

hstojic commented:
I think this should be ready, if tests pass I'll be happy to approve and merge

hstojic committed:
black

Commit:
change notebook wording

the closer or not points sampled on one point at a time is problem dependent, I am using word less accurately since in batch we update our GP not as frequent as one query at a time.

satrialoka commented on trieste/acquisition/function.py:
```diff
@@ -2125,8 +2125,11 @@ def prepare_acquisition_function(
+        try:
+            model.predict_joint(dataset.query_points[0:])
+        except NotImplementedError:
+            raise ValueError(
```
Ah yeah we can try-catch inside the builder, was thinking doing this will be expensive compared to doing try-catch inside the function. my bad.

hstojic replied:
I'm worried about that as well, it could be unnecessarily expensive, even for a single point if its a big dataset - hm, probably we should move it to the acq function itself, it will raise it only on starting to acquire points but its worth the savings - can you please do that?

satrialoka replied:
Sure!

satrialoka commented:
I am happy with the current changes (I add a correction on the notebook text). I think we're ready to merge

Commit:
move try-catch to acq_function

hstojic commented on trieste/acquisition/function.py:
```diff
@@ -2097,3 +2097,71 @@ def _get_min_value_samples(
+    @tf.function
+    def acquisition(x: TensorType) -> TensorType:
+
+        try:
```
@uri-granta is try-except ok inside tf.function?

hstojic replied:
checking this in a builder seems unnecessarily costly

satrialoka committed:
.

hstojic reviewed (approved):
ok, lets merge this in
